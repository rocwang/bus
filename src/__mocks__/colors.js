export default {
  oceanBlue: "#001120",
  green: "#8fc240",
  blue: "#00a3da",
  yellow: "#ffdb1b",
  red: "#da2f2d",
  orange: "#ea8c2e",
  black: "#000",
  grayA: "#333",
  grayB: "#666",
  grayC: "#999",
  grayD: "#ccc",
  white: "#fff",
};
