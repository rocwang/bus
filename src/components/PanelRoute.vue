<template>
  <Panel>
    <template v-slot:icon>
      <RoundIconRoute />
    </template>
    <template v-slot:prefix>
      Route
    </template>
    <template v-slot:title>{{ shortName }}</template>
    <template v-slot:buttons>
      <Buttonizer modifier="icon">
        <button
          v-if="isFavoured"
          @click="del"
          aria-label="Delete from Favourites"
          key="del"
        >
          <IconFullStar />
        </button>
        <button v-else @click="add" aria-label="Add to Favourites" key="add">
          <IconEmptyStar />
        </button>

        <button @click="$router.back()" aria-label="back">
          <IconCross />
        </button>
      </Buttonizer>
    </template>
    <template v-slot:subtitle>
      Stop Code: {{ stopCode }}
    </template>
    <template v-slot:body>
      <ul>
        <li
          v-for="item in tripsWithVehicles"
          :key="item.trip_id + item.departure_time"
        >
          <Trip
            :headSign="item.trip_headsign"
            :departureTime="item.departure_time"
            :isRealTime="!!item.vehicle"
            :occupancyStatus="
              item.vehicle && item.vehicle.occupancy_status
                ? item.vehicle.occupancy_status
                : undefined
            "
          />
        </li>
      </ul>
    </template>
  </Panel>
</template>

<script>
import Panel from "./Panel";
import RoundIconRoute from "./RoundIconRoute";
import Trip from "./Trip";
import IconCross from "./IconCross";
import IconEmptyStar from "./IconEmptyStar";
import IconFullStar from "./IconFullStar";
import Buttonizer from "./Buttonizer";
import {
  actionRemoveFromFavourites$,
  actionAddToFavourites$
} from "../store/actions";

export default {
  name: "PanelRoute",
  components: {
    Buttonizer,
    IconCross,
    IconEmptyStar,
    IconFullStar,
    Panel,
    RoundIconRoute,
    Trip
  },
  props: {
    stopCode: {
      type: String,
      required: true
    },
    shortName: {
      type: String,
      required: true
    },
    trips: {
      type: Array,
      required: true
    },
    vehicles: {
      type: Array,
      required: true
    },
    tripsWithVehicles: {
      type: Array,
      required: true
    },
    favourites: {
      type: Array,
      required: true
    }
  },
  methods: {
    add() {
      actionAddToFavourites$.next({
        stopCode: this.stopCode,
        routeShortName: this.shortName
      });
    },
    del() {
      actionRemoveFromFavourites$.next({
        stopCode: this.stopCode,
        routeShortName: this.shortName
      });
    }
  },
  computed: {
    isFavoured() {
      return !!this.favourites.find(
        ({ stopCode, routeShortName }) =>
          stopCode === this.stopCode && routeShortName === this.shortName
      );
    }
  }
};
</script>
