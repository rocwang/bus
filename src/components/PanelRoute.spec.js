import { mount } from "@vue/test-utils";
import PanelRoute from "./PanelRoute";
import colors from "../__mocks__/colors";
import {
  actionRemoveFromFavourites$,
  actionAddToFavourites$,
} from "../store/actions";

describe("PanelRoute.vue", () => {
  let wrapper;
  let $router;
  const stopCode = "2222";
  const routeShortName = "NX2";

  beforeEach(() => {});

  it("matches the snapshot", () => {
    $router = { back: jest.fn() };
    wrapper = mount(PanelRoute, {
      provide: {
        colors,
      },
      mocks: {
        $router,
      },
      propsData: {
        stopCode,
        routeShortName,
        tripsWithVehicles: [
          {
            trip_id: "NX1",
            departure_time: "17:45:00",
            trip_headsign: "NX1",
          },
          {
            trip_id: "NX2",
            departure_time: "17:05:00",
            trip_headsign: "NX2",
          },
        ],
        favourites: [
          { name: "test1", stopCode: "1111", routeShortName: "NX1" },
          { name: "test2", stopCode: "2222", routeShortName: "NX2" },
        ],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders the stop code", () => {
    expect(wrapper.text()).toMatch(`Stop Code: ${stopCode}`);
  });

  it("renders the route short name", () => {
    expect(wrapper.text()).toMatch(routeShortName);
  });

  it("renders the trips with vehicles", () => {
    expect(wrapper.findAll("li").length).toBe(2);
  });

  it("knows if it is favoured", () => {
    expect(wrapper.vm.isFavoured).toBe(true);
  });

  it("can be added to favourites", () => {
    const expectedPayload = {
      stopCode: "2222",
      routeShortName: "NX2",
    };
    const wrapper = mount(PanelRoute, {
      provide: {
        colors,
      },
      propsData: {
        ...expectedPayload,
      },
    });

    let nextPayload = undefined;
    actionAddToFavourites$.subscribe((payload) => (nextPayload = payload));
    wrapper.find('button[aria-label="Add to Favourites"]').trigger("click");

    expect(nextPayload).toEqual({ stopCode, routeShortName });
  });

  it("can be removed from favourites", () => {
    let nextPayload = undefined;
    actionRemoveFromFavourites$.subscribe((payload) => (nextPayload = payload));
    wrapper
      .find('button[aria-label="Delete from Favourites"]')
      .trigger("click");

    expect(nextPayload).toEqual({ stopCode, routeShortName });
  });

  it("can go back to the last view", () => {
    wrapper.find('button[aria-label="back"]').trigger("click");
    expect($router.back).toBeCalled();
  });
});
