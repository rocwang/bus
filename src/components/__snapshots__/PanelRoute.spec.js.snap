// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelRoute.vue matches the snapshot 1`] = `
<div
  class="root"
>
  <div
    class="head"
  >
    <div
      class="icon"
    >
      <svg
        viewBox="0 0 50 50"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M25 0a25 25 0 1 0 25 24.984C49.991 11.184 38.801 0 25 0"
          fill="#ffdb1b"
        />
         
        <path
          d="M34.375 28.75H28.75a1.88 1.88 0 0 1-1.875-1.875A1.88 1.88 0 0 1 28.75 25h5.625S40 18.73 40 15.625A5.626 5.626 0 0 0 34.375 10a5.626 5.626 0 0 0-5.625 5.625c0 1.494 1.3 3.715 2.654 5.625H28.75a5.634 5.634 0 0 0-5.625 5.625c0 3.1 2.525 5.625 5.625 5.625h5.625a1.88 1.88 0 0 1 1.875 1.875 1.88 1.88 0 0 1-1.875 1.875H20.869A40.453 40.453 0 0 1 18.098 40h16.277c3.1 0 5.625-2.525 5.625-5.625s-2.525-5.625-5.625-5.625zm0-15c1.037 0 1.875.838 1.875 1.875a1.873 1.873 0 0 1-1.875 1.875 1.873 1.873 0 0 1-1.875-1.875c0-1.037.838-1.875 1.875-1.875zM15.625 25A5.626 5.626 0 0 0 10 30.625C10 33.73 15.625 40 15.625 40s5.625-6.27 5.625-9.375A5.626 5.626 0 0 0 15.625 25zm0 7.5a1.873 1.873 0 0 1-1.875-1.875c0-1.037.838-1.875 1.875-1.875s1.875.838 1.875 1.875a1.873 1.873 0 0 1-1.875 1.875z"
          fill="#001120"
        />
      </svg>
    </div>
     
    <h1
      class="title"
    >
      <small
        class="prefix"
      >
        
    Route
  
      </small>
       NX2
    </h1>
     
    <div
      class="buttons"
    >
      <button
        aria-label="Delete from Favourites"
        class="root icon"
      >
        <svg
          viewBox="0 0 30 30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13.39 1.044L9.73 8.8l-8.193 1.248c-1.469.223-2.057 2.115-.992 3.2l5.927 6.034-1.402 8.524c-.252 1.54 1.301 2.695 2.602 1.974L15 25.755l7.329 4.025c1.3.715 2.854-.433 2.602-1.974l-1.402-8.524 5.927-6.035c1.065-1.084.477-2.976-.992-3.199L20.27 8.8l-3.662-7.756a1.763 1.763 0 0 0-3.218 0z"
            fill="#ffdb1b"
          />
        </svg>
      </button>
       
      <button
        aria-label="back"
        class="root icon"
      >
        <svg
          viewBox="0 0 30 30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 15L29 1 15 15 1 1l14 14zm0 0L1 29l14-14 14 14-14-14z"
            fill="none"
            stroke="#ffdb1b"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
     
    <p
      class="subtitle"
    >
      
    Stop Code: 2222
  
    </p>
  </div>
   
  <div
    class="body"
  >
    <ul>
      <li>
        <div
          class="root root"
        >
          <span
            class="headSign"
          >
            
      NX1
      
        (S)
      
          </span>
           
          <span
            class="departureTime"
          >
            
      
    
          </span>
           
          <!---->
           
          <!---->
        </div>
      </li>
      <li>
        <div
          class="root root"
        >
          <span
            class="headSign"
          >
            
      NX2
      
        (S)
      
          </span>
           
          <span
            class="departureTime"
          >
            
      
    
          </span>
           
          <!---->
           
          <!---->
        </div>
      </li>
    </ul>
  </div>
</div>
`;
